File: implement-trie-prefix-tree/obzva.py
'''
풀이
- Trie 자료구조 구현을 따릅니다

Big O
- N: 이전까지 insert한 word의 수
- W: 현재 insert하거나 search하려는 word의 길이
- P: 현재 startsWith하려는 prefix의 길이

- insert
  - Time complexity: O(W)
    - word를 구성하는 모든 문자에 대해 현재 Trie에 저장 여부를 검사합니다
      해시맵 자료구조를 사용하고 있기에 검색과 삽입 모두 constant 시간 복잡도를 가집니다
  - Space complexity: O(W)
    - 최악의 경우 모든 문자를 Trie에 새롭게 추가할 수 있습니다
- search
  - Time complexity: O(W)
    - insert와 같습니다
  - Space complexity: O(1)
- startsWith
  - Time complexity: O(P)
    - search와 같습니다
  - Space complexity: O(1)
'''

class Trie:

    def __init__(self):
        self.root = {}

    def insert(self, word: str) -> None:
        parent = self.root

        for c in word:
            if c not in parent:
                parent[c] = {}
            parent = parent[c]

        parent["word"] = word

    def search(self, word: str) -> bool:
        parent = self.root

        for c in word:
            if c not in parent:
                return False
            parent = parent[c]
        
        return "word" in parent and parent["word"] == word
        

    def startsWith(self, prefix: str) -> bool:
        parent = self.root

        for c in prefix:
            if c not in parent:
                return False
            parent = parent[c]
        
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: search-in-rotated-sorted-array/obzva.go
/*
풀이
- 이진탐색을 두 번 사용하여 풀이할 수 있습니다
  첫번째, pivot index를 찾습니다
  두번째, target을 찾습니다

Big O
- N: 주어진 배열 nums의 길이
- Time complexity: O(logN)
  - 각 이진탐색이 모두 O(logN)의 시간 복잡도를 가집니다
- Space complexity: O(1)
*/

func search(nums []int, target int) int {
	n := len(nums)

	lo, hi := 0, n
	for lo < hi {
		mid := lo + (hi-lo)/2
		if nums[mid] > nums[n-1] {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	pivot := lo
	if pivot == n {
		pivot = 0
	}

	lo, hi = pivot, pivot+n
	for lo < hi {
		mid := lo + (hi-lo)/2
		normalizedMid := mid
		if normalizedMid >= n {
			normalizedMid = mid - n
		}
		if nums[normalizedMid] <= target {
			lo = mid + 1
		} else {
			hi = mid
		}
	}

	if lo > n {
		lo -= n
	}
	if lo-1 < 0 || nums[lo-1] != target {
		return -1
	}
	return lo - 1
}
